elf_load.c

```diff
@@ -311,13 +312,224 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
+ */
+static int elf_load_mmap_filesz_memsz_diff(struct elf_prog *elf_prog, int fd,
+					   GElf_Phdr *phdr __maybe_unused,
+					   uintptr_t vastart, uintptr_t vaend)
+{
+	UK_ASSERT(elf_prog && phdr && vastart && vaend);
+
+	uk_pr_debug("%s: Zeroing 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)(vastart),
+		    (uint64_t)(vaend));
+
+	/* From the last byte contained in filesz to the last
+	 * byte contained in memsz, we can either have:
+	 * 1. a page alignment adjustment, where the end
+	 * address is just PAGE_ALIGN_UP(paddr + filesz), which
+	 * means that our initial call to mmap already read in
+	 * a page for us whose remaining bytes we memset without
+	 * generating a page fault, because
+	 * vaend - vastart < PAGE_SIZE
+	 * ...
+	 */
+	memset((void *)vastart, 0, PAGE_ALIGN_UP(vastart) - vastart);
+
+	if (vaend == PAGE_ALIGN_UP(vastart))
+		return 0;
+
+	/*
+	 * ...
+	 * 2. vaend - vastart >= PAGE_SIZE, which can happen if
+	 * this segment contains a NOBITS section, such as .bss.
+	 * If that is the case, then simply anonymously map this
+	 * remaining area so that we don't waste time memsetting
+	 * it (.bss is quite large).
+	 */
+	vastart = PAGE_ALIGN_UP(vastart);
+	vastart = (uintptr_t)mmap((void *)vastart, vaend - vastart,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+				  fd, 0);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the NOBITS part of phdr at "
+			  "offset %lu\n", phdr->p_offset);
+		return (int)vastart;
+	}
+
+	return 0;
+}
+
+/* Use this to mmap first PT_LOAD */
+static int do_elf_load_fdphdr_0(struct elf_prog *elf_prog,
+				GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart_old, vastart_aligned;
+	uintptr_t vaend_old, vaend_new;
+	uintptr_t vastart, vaend;
+	__sz mmap_len;
+	int rc;
+
+	/* First start/vabase can't be !0 before loading first PT_LOAD */
+	UK_ASSERT(!elf_prog->start && !elf_prog->vabase && phdr);
+
+	if (phdr->p_memsz >= phdr->p_align)
+		mmap_len = phdr->p_memsz + phdr->p_align;
+	else
+		mmap_len = phdr->p_align * 2;
+
+	vastart = (uintptr_t)mmap(NULL, mmap_len,
+				  PROT_EXEC | PROT_READ | PROT_WRITE,
+				  MAP_PRIVATE,
+				  fd, phdr->p_offset);
+	if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+		uk_pr_err("Failed to mmap the phdr at offset %lu\n",
+			  phdr->p_offset);
+		return (int)vastart;
+	}
+
+	uk_pr_debug("%s: Memory mapped 0x%"PRIx64" - 0x%"PRIx64" to 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)phdr->p_offset,
+		    (uint64_t)phdr->p_filesz,
+		    (uint64_t)vastart,
+		    (uint64_t)vastart + (uint64_t)phdr->p_filesz);
+
+	vastart_aligned = ALIGN_UP(vastart, phdr->p_align);
+	vastart_old = vastart;
+
+	/* Force remap with MAP_FIXED */
+	if (vastart_aligned != vastart) {
+		vastart = (uintptr_t)mmap((void *)vastart_aligned,
+					  phdr->p_filesz,
+					  PROT_EXEC | PROT_READ | PROT_WRITE,
+					  MAP_PRIVATE | MAP_FIXED,
+					  fd, phdr->p_offset);
```

Comment:
While we test this PR with dynamic applications, we are actually testing with the `ld-linux.so` loader. Does loader know where we put the dynamic ELF? Or the loader will load the application by itself?

When I got some static-linked apps for testing, they crushed immediately.

Reply:
Should our `mmap` implement check for alignment and reject non-page-aligned offsets? Or it is a feature?

Reply:
> What will happen if p_offset is not page-aligned...

Hmm, yes, I wrongly assumed that, because it is the first segment, its p_offset is always aligned.

> Should our mmap implement check for alignment and reject non-page-aligned offsets? Or it is a feature?

Good point, I will bring up a discussion about this in a related meeting :+1:.

> Does loader know where we put the dynamic ELF? Or the loader will load the application by itself?

The loader loads the linker and the app and passes the base address among other things as arguments to the linker before passing execution to the linker.

> When I got some static-linked apps for testing, they crushed immediately.

Does this also happen when not using the mmap version (`CONFIG_LIBPOSIX_MMAP=y`). What about the staging branch? :thinking: I can't really get it to reproduce.

Reply:
> Hmm, yes, I wrongly assumed that, because it is the first segment, its p_offset is always aligned.

Maybe we can continue to assume that, just add a `UK_ASSERT`. I also haven't seen any program violating this.

> The loader loads the linker and the app and passes the base address among other things as arguments to the linker before passing execution to the linker.

Thanks, I now have understood how interpreter works...

> Does this also happen when not using the mmap version (`CONFIG_LIBPOSIX_MMAP=y`). What about the staging branch? 🤔 I can't really get it to reproduce.

It was because of another compatibility bug, not related to this, sorry I forgot there exists such incompatibilities.